nifi-nar-bundles/nifi-avro-bundle/nifi-avro-processors/src/main/java/org/apache/nifi/processors/avro/ConvertAvroToJSON.java

```diff
@@ -116,51 +124,59 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
+                         final OutputStream out = new BufferedOutputStream(rawOut);
+                         final DataFileStream<GenericRecord> reader = new DataFileStream<>(in, new GenericDatumReader<GenericRecord>())) {
 
                         final GenericData genericData = GenericData.get();
 
-                        if (reader.hasNext() == false ) {
-                            out.write(EMPTY_JSON_OBJECT);
-                            return;
+                        int recordCount = 0;
+                        GenericRecord currRecord = null;
+                        if (reader.hasNext()) {
+                            currRecord = reader.next();
+                            recordCount++;
                         }
-                        int recordCount = 1;
-                        GenericRecord reuse = reader.next();
-                        // Only open container if more than one record
-                        if(reader.hasNext() && containerOption.equals(CONTAINER_ARRAY)){
+
+                        // Open container if desired output is an array format and there are are multiple records or
+                        // if configured to wrap single record
+                        if (reader.hasNext() && useContainer || wrapSingleRecord) {
                             out.write('[');
                         }
-                        out.write(genericData.toString(reuse).getBytes(StandardCharsets.UTF_8));
+
+                        // Determine the initial output record, inclusive if we should have an empty set of Avro records
+                        final byte[] outputBytes = (currRecord == null) ? EMPTY_JSON_OBJECT : genericData.toString(currRecord).getBytes(StandardCharsets.UTF_8);
+                        out.write(outputBytes);
 
                         while (reader.hasNext()) {
-                            if (containerOption.equals(CONTAINER_ARRAY)) {
+                            if (useContainer) {
                                 out.write(',');
                             } else {
                                 out.write('\n');
                             }
 
-                            reuse = reader.next(reuse);
-                            out.write(genericData.toString(reuse).getBytes(StandardCharsets.UTF_8));
+                            currRecord = reader.next(currRecord);
+                            out.write(genericData.toString(currRecord).getBytes(StandardCharsets.UTF_8));
                             recordCount++;
                         }
 
-                        // Only close container if more than one record
-                        if (recordCount > 1 && containerOption.equals(CONTAINER_ARRAY)) {
+                        // Close container if desired output is an array format and there are multiple records or if
+                        // configured to wrap a single record
+                        if (recordCount > 1 && useContainer || wrapSingleRecord) {
```

Comment:
I think it is correct.

useContainer only applies to the case where there is more than one record.  Those cases where there are zero or one are driven by wrapSingleRecord which incorporates whether or not the container should be used.  Happy to rework this (and the grouping) if I can get some better clarity.   Could certainly alleviate the line where it is defined earlier.

wrapSingleRecord is defined as:
final boolean wrapSingleRecord = context.getProperty(WRAP_SINGLE_RECORD).asBoolean() && useContainer;